sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/ImplementSerializationMethodsCorrectlyBase.cs

```diff
@@ -0,0 +1,98 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2018 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules
+{
+    public abstract class ImplementSerializationMethodsCorrectlyBase : SonarDiagnosticAnalyzer
+    {
+        internal const string DiagnosticId = "S3927";
+        protected const string MessageFormat = "Make this method {0}.";
+        private const string problemMakePrivateText = "'private'";
+        private const string problemReturnVoidText = "return 'void'";
```

Comment:
A more precise message would be "Make this function a sub" or "This method should not return a value" or something like this...